design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,293 @@
+```
+
+#### Secrets Encryption
+
+The JSON schema can contain a property called `non-secrets` which expects an array of JSON schema property names. The properties specified in this array will not be encrypted in the database, along with the JSON structure. The unspecified attributes, will by default get encrypted. This allows the protection of sensitive data while also allowing the non-sensitive values to be searchable. Meaning that the variable will not be encrypted as a whole, but only the sensitive attributes in the JSON structure.
+
+For the example above, the `mysql-db-creds` variable will be saved in the database as follows:
+
+```json
+{
+  "username": "my-user",
+  "password": "U2FsdGVkX19Ji6JpgnVHW3V47OtMJwuKi1Yf9nc0aP5QcuzdnIrpzZ2zMC90f24g",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+Before the variable value is returned to the client, the secret attributes are decrypted so that the given example above, would look like this:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+#### Built-in Schemas
+
+To simplify the user experience, Conjur can come with prdefined JSON schemas for the most common secrets use cases. The list of predefined schmeas should include:
+
+- Database secrets - Oracle, MSSQL, MySQL, Postgres, MariaDB, DB2.
+- Cloud access keys - AWS, Azure, GCP.
+- X509 certificates.
+- JWTs
+
+Write about the advantages of atomicity, minimal changes in the database to implement this and so on.
+
+#### Automatic Permission Granting to Built-in Schemas
+
+In order to provide access for any role to the built-in schemas, a new built-in group should be introduced: `conjur/all`. All roles in the Conjur account should be added to this group automatically, thus keeping this group always up to date with all the roles that exist in the account.
+
+Usage example:
+
+```yaml
+- !permit
+  role: !group /conjur/all
+  privileges: [ read,execute ]
+  resource: !variable my-var
+```
+
+## Main Advantages and Expected Value From This Proposal
+
+- We leverage a standard way to enforce the content of the variable, with a well known JSON schema.
+- Tightly coupled values, such as username and password, are updated together in a single transaction. This will prevent momentary inconsistency in which each value was updated independently, one after the other.
+- The user will not have to look in the documentation or to understand the required variables, their structure or allowed input. Conjur will provide this feedback when an update attempt is made.
+- The feature does not require changes in the database schma, only in new built-in content.
+- The returned variable is in a JSON structure, same as the rest of our APIs responses.
+
+## Affected Areas
+
+- The new functionality will be developed in the Conjur server and the policy parser.
+- No new APIs are needed, since all the functionality will be given as part of the policy loading.
```

Comment:
I think we will need an api to fetch all the variables ids under a specific schema? one use case i can think of is if you will want to delete a schema

Reply:
I added a section to discuss this point, thanks for asking about that!
I think the solution should be cascading removal of the annotations from the variables that had references to the deleted JSON schema. It should be also very easy to find them in the `annotations` table in the database and delete them.

Reply:
I think API fetching is a great idea and it is very good it is elaborated in design